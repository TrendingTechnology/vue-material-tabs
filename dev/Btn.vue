<template>
  <button class="btn" @click="$emit('click')">
    <slot />
  </button>
</template>

<style scoped>
.btn {
  border: none;
  align-items: center;
  border-radius: 4px;
  display: inline-flex;
  flex: 0 0 auto;
  font-weight: 500;
  letter-spacing: 0.0892857143em;
  justify-content: center;
  outline: 0;
  position: relative;
  text-decoration: none;
  text-indent: 0.0892857143em;
  text-transform: uppercase;
  vertical-align: middle;
  white-space: nowrap;
  background-color: #1867c0;
  cursor: pointer;
  padding: 0.7rem 2rem;
  color: white;
  transition: background 0.28s;
}

.btn:hover {
  background: #1e7be6;
}
</style>
