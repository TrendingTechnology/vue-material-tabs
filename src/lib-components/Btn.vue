<template>
  <button class="btn" @click="$emit('click')">
    <svg
      xmlns="http://www.w3.org/2000/svg"
      viewBox="0 0 24 24"
      role="img"
      aria-hidden="true"
    >
      <path d="M15.41,16.58L10.83,12L15.41,7.41L14,6L8,12L14,18L15.41,16.58Z" />
    </svg>
  </button>
</template>

<script>
export default {
  name: "Btn",
};
</script>

<style scoped>
.btn {
  outline: none;
  border: none;
  background: transparent;
  cursor: pointer;
}

.btn svg {
  height: 24px;
  width: 24px;
}

.btn:disabled svg {
  fill: rgb(214, 213, 213);
  cursor: default;
}
</style>
