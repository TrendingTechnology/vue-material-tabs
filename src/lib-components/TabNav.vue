<template>
  <div v-resize="resizable" :class="classes">
    <div class="tab-pagination__prev">
      <Btn
        v-if="pagination.has"
        :disabled="!paginateIndicator.previous"
        @click="paginationHandler('previous')"
      />
    </div>
    <nav ref="nav" class="tab__nav">
      <ul ref="navItems" class="tab__nav__items" :style="styles">
        <li
          v-ripple="ripple"
          v-for="(navItem, index) in navItems"
          :key="`tab-item-${index}`"
          :ref="navItem.model"
          class="tab__nav__item"
          :class="{
            active: navItem.model === tabItemActive.model,
            disabled: navItem.disabled,
          }"
          @click.prevent="select(navItem)"
        >
          <VNode v-if="navItem.nameSlot" :node="navItem.nameSlot" />
          <span v-else>
            {{ navItem.name }}
          </span>
        </li>
        <hr v-if="navSlider" class="tab__slider" />
      </ul>
    </nav>
    <div class="tab-pagination__next">
      <Btn
        v-if="pagination.has"
        :disabled="!paginateIndicator.next"
        @click="paginationHandler('next')"
      />
    </div>
  </div>
</template>

<script>
import ripple from "./ripple/index";
import resize from "./resize/index";
import Btn from "./Btn";

export default {
  components: {
    Btn,
    VNode: {
      functional: true,
      render: (h, ctx) => h("span", [ctx.props.node]),
    },
  },

  directives: { ripple, resize },

  props: {
    vertical: Boolean,
    navSlider: Boolean,
    ripple: Boolean,
    navAuto: Boolean,
    navItems: {
      type: Array,
      required: true,
    },
    tabItemActive: {
      type: Object,
      default: () => {},
    },
  },

  data: () => ({
    pagination: {
      has: false,
      maxOffset: 0,
      minOffset: 0,
      translate: 0,
      offset: 0,
    },
  }),

  computed: {
    classes() {
      return {
        tab__pagination: true,
        "tab__pagination--vertical": this.vertical,
        "tabs__nav--auto": this.navAuto && !this.vertical,
      };
    },

    styles() {
      return {
        transform: `translate${this.vertical ? "Y" : "X"}(-${
          this.pagination.translate
        }px)`,
      };
    },

    paginateIndicator() {
      return {
        next: this.pagination.translate < this.pagination.maxOffset,
        previous: this.pagination.translate > 0,
      };
    },

    orientation() {
      return this.vertical ? "portrait" : "landscape";
    },
  },

  watch: {
    vertical: ["sliderHandler", "setPaginationOffset"],
    tabItemActive: ["sliderHandler", "paginationBySomething"],
  },

  mounted() {
    this.setPaginationOffset();
  },

  methods: {
    select(navItem) {
      this.$emit("select", {
        tabItem: navItem,
        byUser: true,
      });
    },

    async sliderHandler() {
      await this.$nextTick();

      const navActiveItemElement = this.$refs?.[this.tabItemActive.model]?.[0];
      const navItemsElement = this.$refs?.navItems;
      if (navActiveItemElement && navItemsElement) {
        const {
          width,
          height,
          left: navActiveLeft,
          top: navActiveTop,
        } = navActiveItemElement.getBoundingClientRect();

        const {
          left: navItemsLeft,
          top: navItemsTop,
        } = navItemsElement.getBoundingClientRect();

        const { children } = navItemsElement;
        const sliderElement = children[children.length - 1];
        sliderElement.removeAttribute("style");

        Object.assign(
          sliderElement.style,
          {
            portrait: {
              height: `${height}px`,
              top: `${navActiveTop - navItemsTop}px`,
            },
            landscape: {
              width: `${width}px`,
              left: `${navActiveLeft - navItemsLeft}px`,
            },
          }[this.orientation]
        );
      }
    },

    async setPaginationOffset() {
      await this.$nextTick();

      const navElement = this.$refs?.nav;
      if (!navElement) return;

      const {
        width: navItemsWidth,
        height: navItemsHeight,
      } = navElement.children[0].getBoundingClientRect();

      const {
        width: navWidth,
        height: navHeight,
      } = navElement.getBoundingClientRect();

      const paginationFactory = (has, maxOffset, minOffset) => ({
        has,
        maxOffset,
        minOffset,
        offset: minOffset,
      });

      Object.assign(
        this.pagination,
        {
          portrait: paginationFactory(
            navItemsHeight > navHeight,
            navItemsHeight - navHeight,
            navHeight
          ),
          landscape: paginationFactory(
            navItemsWidth > navWidth,
            navItemsWidth - navWidth,
            navWidth
          ),
        }[this.orientation]
      );
    },

    paginationHandler(type) {
      const { maxOffset, offset, translate, minOffset } = this.pagination;
      if (type === "previous" && this.paginateIndicator.previous) {
        if (offset <= minOffset) {
          this.pagination.offset = minOffset;
        }

        if (translate - offset < offset) {
          this.pagination.translate = 0;
          return;
        }

        this.pagination.translate = translate - offset;
      }

      if (type === "next" && this.paginateIndicator.next) {
        if (translate + offset > maxOffset) {
          this.pagination.offset = maxOffset - translate;
        }
        this.pagination.translate = translate + this.pagination.offset;
      }
    },

    paginationBySomething({ model }) {
      const navActiveItemElement = this.$refs?.[model]?.[0];
      const navElement = this.$refs?.nav;

      if (navActiveItemElement && navElement) {
        let toTranslate = this.pagination.translate;

        const {
          right: navActiveRight,
          left: navActiveLeft,
        } = navActiveItemElement.getBoundingClientRect();

        const {
          right: navRight,
          left: navLeft,
        } = navElement.getBoundingClientRect();

        if (navActiveRight > navRight) {
          toTranslate = toTranslate + (navActiveRight - navRight);
        }

        if (navActiveLeft < navLeft) {
          toTranslate = toTranslate - (navLeft - navActiveLeft);
        }

        this.pagination.translate = Math.floor(toTranslate);
      }
    },

    resizable() {
      this.setPaginationOffset();
      this.sliderHandler();
    },
  },
};
</script>

<style scoped>
.tab__pagination {
  display: flex;
  justify-content: space-between;
  align-items: center;
  vertical-align: middle;
  max-width: 100%;
  flex: 0 1 auto;
  position: relative;
  contain: content;
}

.tab__pagination .tab-pagination__prev,
.tab-pagination__next {
  flex: 1 40px;
  min-width: 40px;
}

.tab-pagination__next >>> .btn svg {
  transform: rotate(180deg);
}

.tab__nav {
  position: relative;
  display: flex;
  overflow: hidden;
  flex: 1 100%;
}

.tab__nav__items {
  display: flex;
  margin: 0;
  padding: 0;
  flex: 1 auto;
  transition: 0.3s cubic-bezier(0.25, 0.8, 0.5, 1);
  height: 100%;
}

.tab__nav__items .tab__nav__item {
  list-style: none;
  text-align: center;
  cursor: pointer;
  padding: 0.9rem 1rem;
  letter-spacing: 0.0892857143em;
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  color: gray;
  text-transform: uppercase;
  font-size: 0.875rem;
  font-weight: 500;
  white-space: normal;
  transition: background 0.1s ease;
  position: relative;
  overflow: hidden;
  min-width: 90px;
  max-width: 360px;
  user-select: none;
}

.tab__nav__items .tab__nav__item:hover:not(.disabled) {
  background: #faf9f9;
}

.tab__nav__items .active {
  color: black;
  color: #1867c0;
}

.tab__nav__items .active:hover {
  background: #1b7ef01c !important;
}

.tab__nav__items .disabled {
  background: #f3f2f2;
}

.tab__slider {
  height: 2px;
  width: 2px;
  background: #1867c0;
  border: none;
  margin: 0;
  padding: 0;
  bottom: 0;
  position: absolute;
  transition: left 0.3s cubic-bezier(0.25, 0.8, 0.5, 1),
    top 0.3s cubic-bezier(0.25, 0.8, 0.5, 1);
}

/* Vertical */
.tab__pagination--vertical {
  flex-direction: column;
}

.tab__pagination--vertical .tab__nav__items {
  flex-direction: column;
  flex: 1 auto;
  position: relative;
}

.tab__pagination--vertical >>> .tab-pagination__prev svg {
  transform: rotate(90deg);
}

.tab__pagination--vertical >>> .tab-pagination__next svg {
  transform: rotate(270deg);
}

.tab__pagination--vertical .tab__nav__item {
  justify-content: left;
  padding-top: 1.6rem;
  padding-bottom: 1.6rem;
}

.tabs--dark .tab__nav__item:hover {
  background: #2f3236;
}

/* Nav auto */
.tabs__nav--auto .tab__nav__item {
  flex: 1 auto;
}
</style>

<style>
.ripple {
  background-color: #1866c04d;
  border-radius: 50%;
  position: absolute;
  transform: scale(0);
  animation: ripple 0.6s linear;
  z-index: 2;
}

@keyframes ripple {
  to {
    transform: scale(4);
    opacity: 0;
  }
}
</style>
