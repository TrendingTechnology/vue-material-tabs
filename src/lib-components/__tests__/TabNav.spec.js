import { mount, createLocalVue } from "@vue/test-utils";
import TabNav from "../TabNav.vue";
import Btn from "../Btn.vue";

const localVue = createLocalVue();
localVue.component(Btn.name, Btn);

describe("TabNav.vue", () => {
  const navItems = [
    {
      name: "Test",
      model: "123",
      disabled: false,
    },
  ];

  const mountFunction = (options = {}) => {
    return mount(TabNav, {
      ...options,
      localVue,
      propsData: {
        navItems,
        tabItemActive: navItems[0],
      },
    });
  };

  it("Should mount the component", () => {
    const wrapper = mountFunction();
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
});
