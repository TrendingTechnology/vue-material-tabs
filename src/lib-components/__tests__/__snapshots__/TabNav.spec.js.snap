// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabNav.vue Should mount the component 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;
